### PR TITLE
[RDY]Save to OS temp directory if the intended file can't be written to

### LINKED ISSUE #470: Configuration and save games reset on Mac OS X
_From [forward....@gmail.com](https://code.google.com/u/102141427359698628586/) on July 08, 2011 15:45:59_

What steps will reproduce the problem? Unfortunately nothing seemed to go wrong while playing, the game exited as normal. When starting up an error during utility.loadfile line 123 occurred (result is nil).

Looking at the code this occurs when the config.txt file exists but it has no data in it.

After finding the config file it was zero bytes and had no data.

Deleting the config file caused it to be recreated from scratch when corsix started up, however the previous active save game file and the autosave file were both zero bytes (we were at level 11!). What is the expected output? What do you see instead? I expect the save game files to remain in the correct state.
My suggestion would be that corsix backs up the file it is about to overwrite and then performs the overwrite. This will at least allow a manual restoration of game state.

What version of CorsixTH are you using (e.g. "Beta 2", "SVN r123 ")?
Beta 7. What operating system / compile settings are you using? Mac OS X

What level was this on (e.g. "Demo level", "Full game level 12")?
Full game level 11. Attach a saved game file if you have one and it is relevant. Please provide any additional information below.

_Original issue: http://code.google.com/p/corsix-th/issues/detail?id=928_

### PR DESCRIPTION
**Pull request by tobylane**
*Fixes #470*

**Describe what the proposed change does**
- On failing to write to the intended save game file it saves to a temp file and tells the user

### CODE REVIEW

File: CorsixTH/Lua/persistance.lua [hunk 1]
```diff
@@ -255,7 +255,14 @@ end
 --!param filename (string) Path of the file to write.
 function SaveGameFile(filename)
   local data = SaveGame()
-  local f = assert(io.open(filename, "wb"))
+  local f, err = io.open(filename, "wb")
+  if err then
+    local file = os.tmpname()
+    f = io.open(file, "wb")
+    TheApp.ui:addWindow(UIInformation(TheApp.ui,
+        {"Saved " .. filename .. " to " .. file, "Error:", err}))
```
Comment (TheCycoONE) [inline CorsixTH/Lua/persistance.lua:263]: This is (potentially) user facing. Should use the language file and make a user friendly ish message.

Comment (tobylane) [reply CorsixTH/Lua/persistance.lua]: Done, added _S.errors.save_to_tmp.

**Comment by tobylane** [comment]
Made it generic and applied it to the other main io.writes. A few minor ones were left out, the first two here https://grep.app/search?q=%22wb%3F%22&regexp=true&filter[repo][0]=CorsixTH/CorsixTH&filter[lang][0]=Lua&filter[path][0]=CorsixTH/Lua/

Not sure about the App:writeToFileOrTmp name.

Apologies for the multiple pushes.

**Comment by tobylane** [comment]
Rebased to pass tests.

**Comment by tobylane** [comment]
I can't re run the AppVeyor test but it's a Lua only change. Ready.

**Pull request merged**